qualtran/bloqs/multi_control_multi_target_pauli.py

```diff
@@ -137,6 +140,160 @@ def _has_unitary_(self) -> bool:
         return True
 
 
+@frozen
+class MultiControlX(Bloq):
+    r"""Implements multi-control, single-target X gate as a bloq using $n-1$ clean ancillas.
+
+    References:
+        [Constructing Large Controlled Nots]
+        (https://algassert.com/circuits/2015/06/05/Constructing-Large-Controlled-Nots.html)
+    """
+
+    cvs: Tuple[int, ...] = field(converter=lambda v: (v,) if isinstance(v, int) else tuple(v))
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        if len(self.cvs) > 0:
+            return Signature([Register('ctrl', bitsize=len(self.cvs)), Register('x', bitsize=1)])
+        else:
+            return Signature([Register('x', bitsize=1)])
+
+    def on_classical_vals(
+        self, x: 'ClassicalValT', **vars: 'ClassicalValT'
+    ) -> Dict[str, 'ClassicalValT']:
+        if len(self.cvs) > 0:
+            ctrl = vars['ctrl']
+        else:
+            return {'x': (x + 1) % 2}
+
+        ctrls = bits_to_ints(self.cvs)
+        if ctrls == ctrl:
+            x = (x + 1) % 2
+
+        return {'ctrl': ctrl, 'x': x}
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', x: SoquetT, **regs: SoquetT
+    ) -> Dict[str, 'SoquetT']:
+
+        # n = number of controls in the bloq.
+        n = len(self.cvs)
+
+        # Base case 0: XGate()
+        if n == 0:
+            x = bb.add(XGate(), q=x)
+            return {'x': x}
+
+        # Base case 1: CNOT()
+        if n == 1:
+            ctrl = regs['ctrl']
+            # Allows for 0-controlled implementations.
+            if self.cvs[0] == 0:
+                ctrl = bb.add(XGate(), q=ctrl)
+            ctrl, x = bb.add(CNOT(), ctrl=ctrl, target=x)
+            if self.cvs[0] == 0:
+                ctrl = bb.add(XGate(), q=ctrl)
+            return {'ctrl': ctrl, 'x': x}
+
+        # Base case 2: Toffoli()
+        if n == 2:
+            ctrls = regs['ctrl']
+            ctrls_split = bb.split(ctrls)
+
+            # Allows for 0-controlled implementations.
+            for i in range(len(self.cvs)):
+                if self.cvs[i] == 0:
+                    ctrls_split[i] = bb.add(XGate(), q=ctrls_split[i])
+
+            ctrl = [ctrls_split[0], ctrls_split[1]]
+            ctrl, x = bb.add(Toffoli(), ctrl=ctrl, target=x)
```

Comment:
`ctrls_split` is already a length-2 1d array, right? you can just pass that in to `bb.add(... ctrl=ctrls_split)`